### ISSUE TITLE
Test and document how to produce typed NA/missing values in Python. 

### ISSUE THREAD

**Issue by zkamvar** [hubverse-org/hubDocs#198]
As mentioned in https://github.com/reichlab/variant-nowcast-hub/pull/116#issuecomment-2427130801:

> If we coerce to `pd.NA` (or `None`), the corresponding dtype that the validation tool receives is `vctrs_unspecified`, rather than `chr`. Similarly, using np.nan to encode gives a "double" (numeric) data type, instead of char. Seems like an issue to fix on the backend.

This might be addressed partially in https://github.com/hubverse-org/schemas/issues/109, but I wonder if it's possible to catch `vctrs_unspecified` and convert them to characters since we know those are going to always be missing values.

**Comment by zkamvar** [comment]
I did a little experiment and it seems like we can either convert to `float` or `string`, but integer is not possible because it throws the error: `Cannot convert non-finite values (NA or inf) to integer`. 

```python
import pandas as pd
import numpy as np
import math as math

# Define the data as a list of dictionaries
data = [
    {"NA": pd.NA, "None": None, "math.nan": math.nan, "np.nan": np.nan, "floatnan": float('nan')},
    {"NA": pd.NA, "None": None, "math.nan": math.nan, "np.nan": np.nan, "floatnan": float('nan')},
]

# Convert the list of dictionaries to a pandas DataFrame
df = pd.DataFrame(data)

df["NA"] = df["NA"].astype("string")
df["None"] = df["None"].astype("string")
# Save the DataFrame
df.to_csv("string.csv", index = False, na_rep = "NA")
df.to_parquet("string.parquet")

df["NA"] = df["NA"].astype("float")
df["None"] = df["None"].astype("float")
# Save the DataFrame
df.to_csv("float.csv", index = False, na_rep = "NA")
df.to_parquet("float.parquet")
```

**Comment by annakrystalli** [comment]
Nice one @zkamvar ! I was wondering actually whether we should put together some example files in an example hub where we actually write some of these out and can then test whether they:
- correctly validate
- can be successfully accessed and filtered on
- other downstream tools can correctly parse in both R and python (e.g. polars see @elray1's comment)

So far the results of your experiment make me wonder if we should restrict output type ID columns to be one of `"double"` or `"character"` for all output types in the schema (currently we allow "double"`, `"character"` or `"integer"`.

**Comment by annakrystalli** [comment]
I think this mini example hub should follow v4 schema and probably live in `hubUtils` `inst` to be available more broadly for examples and testing.

**Comment by zkamvar** [comment]
> So far the results of your experiment make me wonder if we should restrict output type ID columns to be one of `"double"` or `"character"` for all output types in the schema (currently we allow `"double"`, `"character"` or `"integer"`.

My mistake. I was using `astype("int64")` instead of `astype("Int64")`. I updated the example. 

I found that the output files above are all able to produce missing data that is interoperable with R and Python. 

> It seems the experimental `pd.NA` is the closest to Rs and likely what we want to promote 

I would avoid trying to design the data structure in a way that makes it more compatible with R at the expense of natural semantics in other languages. Instead, we should focus on the data formats themselves and how they represent missing data. 

For JSON and Arrow, that's a single `null` value. For CSV, it's more nuanced (Python, Julia, and R can all understand that an empty cell---two juxtaposed commas---represents missing data, but R established the explicit convention of labelling missing data as `NA`, which prevents an empty string `, ,` from being misinterpreted), but `NA` is a safe standard as it's in all the documentation.

**Comment by elray1** [comment]
I used the following code to read in a csv format file and create data frames in both polars and pandas which passed validations.  This is working with the example submission csv file in this folder: https://github.com/elray1/FluSight-forecast-hub/tree/main/auxiliary-data

```
import pandas as pd
import polars as pl

submission_pl = pl.read_csv(
    "auxiliary-data/2024-11-16-example-submission.csv",
    ignore_errors=True
)

submission_pl = submission_pl.with_columns(
    pl.col("reference_date").str.to_date("%Y-%m-%d"),
    pl.col("target_end_date").replace("NA", None).str.to_date("%Y-%m-%d")
)

submission_pd.write_parquet("auxiliary-data/2024-11-16-example-submission.parquet")



submission_pd = pd.read_csv(
    "auxiliary-data/2024-11-16-example-submission.csv"
)

for col in ["reference_date", "target_end_date"]:
    submission_pd[col] = pd.to_datetime(submission_pd[col]).dt.date

submission_pd["horizon"] = submission_pd["horizon"].astype('Int64')

submission_pd.to_parquet("auxiliary-data/2024-11-16-example-submission.parquet")
```

**Comment by bsweger** [comment]
> I did a little experiment and it seems like python tolerates `None` fairly well in the conversions.
> 

Nice investigation! Chiming in to say that @zkamvar helped me frame the problem like this:

- _How_ missing data is encoded in a model output file is a concern of the language being used to generate it
- The concern of modelers is not really the representation of missing data--it's the data type of the column that contains it. 

_If_ that framing is correct, the updated docs and other artifacts should focus on the second item 

> The only sticking thing is that pandas with a numpy_nullable dtype backend does not like to read integer columns with missing data. Instead, it will silently convert those columns to floats.

That's a PITA and also a good reason to focus people on getting their column data types correct

**Comment by zkamvar** [comment]
I believe this issue was closed by #197, which went through several rounds of feedback that addressed the comments of this thread. It encourages modelers to focus on the data type of the column and defines how missing data is represented in the model output (arrow: null, CSV: NA or blank cell. 

WRT to the testing component, I feel that this is resolved because of the two points brought up in https://github.com/hubverse-org/hubDocs/issues/198#issuecomment-2448127804:

> * _How_ missing data is encoded in a model output file is a concern of the language being used to generate it
> * The concern of modelers is not really the representation of missing data--it's the data type of the column that contains it.

Which makes sense because modelers are free to construct their models in any way, shape, or form, as long as their outputs can be presented in a file that matches at the hub's specifications (i.e. it doesn't matter if models are created in Julia, Rust, Excel, or divination; it's all good as long as they can produce a CSV file).

Moreover, the demonstrations in https://github.com/hubverse-org/hubDocs/issues/198#issuecomment-2430287818 and https://github.com/hubverse-org/hubDocs/issues/198#issuecomment-2447157669 sufficiently show that this is possible.
